### PR TITLE
fix: check whether the stale DNS plugin needs to do a force topology refresh

### PR DESCRIPTION
**Pull request by karenc-bq**
### Summary

fix: check whether the stale DNS plugin needs to do a force topology refresh

### Additional Reviewers

@sergiyvamz

### CODE REVIEW

File: wrapper/src/main/java/software/amazon/jdbc/plugin/staledns/AuroraStaleDnsHelper.java [hunk 1]
```diff
@@ -74,7 +77,12 @@ public Connection getVerifiedConnection(
       return conn;
     }
 
-    this.pluginService.refreshHostList(conn);
+    if (isReadOnly(conn, query)) {
+      this.pluginService.forceRefreshHostList(conn);
```
Comment (sergiyvamz) [inline wrapper/src/main/java/software/amazon/jdbc/plugin/staledns/AuroraStaleDnsHelper.java:81]: Let's add descriptive comment why force topology refresh is needed here.

File: wrapper/src/main/java/software/amazon/jdbc/plugin/staledns/AuroraStaleDnsHelper.java [hunk 2]
```diff
@@ -153,4 +161,16 @@ private HostSpec getWriter() {
...
   }
+
+  private boolean isReadOnly(Connection connection, final String query) {
+    try (final Statement statement = connection.createStatement();
+        ResultSet rs = statement.executeQuery(query)) {
+      while (rs.next()) {
+        return rs.getBoolean(AuroraStaleDnsPlugin.IS_READER_COLUMN);
+      }
+    } catch (SQLException e) {
+      return false;
```
Comment (sergiyvamz) [inline wrapper/src/main/java/software/amazon/jdbc/plugin/staledns/AuroraStaleDnsHelper.java:172]: I'd separate sql syntax exception from others. Syntax exception means that the connection is not to an Aurora cluster and it should re-thrown. Communication-related exceptions probably require a retry. I wouldn't simply return false in such cases.

File: wrapper/src/main/java/software/amazon/jdbc/plugin/staledns/AuroraStaleDnsPlugin.java [hunk 3]
```diff
@@ -55,6 +55,16 @@ public class AuroraStaleDnsPlugin extends AbstractConnectionPlugin {
       Arrays.asList("connect", "initHostProvider", "notifyNodeListChanged", "*")));
   private final PluginService pluginService;
   private final AuroraStaleDnsHelper helper;
+  private static final String POSTGRESQL_READONLY_QUERY = "SELECT rt.server_id, CASE\n"
```
Comment (sergiyvamz) [inline wrapper/src/main/java/software/amazon/jdbc/plugin/staledns/AuroraStaleDnsPlugin.java:58]: Will it work for Aurora Mysql?

File: wrapper/src/main/java/software/amazon/jdbc/plugin/staledns/AuroraStaleDnsHelper.java (hunk 2 above)
Comment (davecramer) [reply wrapper/src/main/java/software/amazon/jdbc/plugin/staledns/AuroraStaleDnsHelper.java:172]: This should probably be an if statement not a while statement

File: wrapper/src/main/java/software/amazon/jdbc/plugin/staledns/AuroraStaleDnsHelper.java [hunk 4]
```diff
@@ -153,4 +167,25 @@ private HostSpec getWriter() {
...
+          ResultSet rs = statement.executeQuery(query)) {
+        if (rs.next()) {
+          return rs.getBoolean(AuroraStaleDnsPlugin.IS_READER_COLUMN);
+        }
+      } catch (SQLSyntaxErrorException e) {
+        throw e;
+      } catch (SQLException e) {
+
+        // Return false if the SQLException is not a network error. Otherwise, retry.
+        if (!SqlState.isConnectionError(e)) {
```
Comment (sergiyvamz) [inline wrapper/src/main/java/software/amazon/jdbc/plugin/staledns/AuroraStaleDnsHelper.java:184]: Should we use our custom exception handler?

File: wrapper/src/main/java/software/amazon/jdbc/plugin/staledns/AuroraStaleDnsPlugin.java (hunk 3 above)
Comment (sergiyvamz) [inline wrapper/src/main/java/software/amazon/jdbc/plugin/staledns/AuroraStaleDnsPlugin.java:58]: Would you consider to move these queries to AuroraStaleDnsHelper? Then getVerifiedConnection() can accept driver protocol and decide what query to run.

File: wrapper/src/main/java/software/amazon/jdbc/plugin/staledns/AuroraStaleDnsHelper.java [hunk 5]
```diff
@@ -48,16 +48,29 @@ public class AuroraStaleDnsHelper {
...
   private static final int RETRIES = 3;
 
+  private static final String POSTGRESQL_READONLY_QUERY = "SELECT rt.server_id, CASE\n"
+      + "           WHEN 'MASTER_SESSION_ID' = rt.session_id THEN 'false'\n"
+      + "           ELSE 'true'\n"
+      + "           END AS is_reader\n"
+      + "FROM aurora_replica_status() rt,\n"
+      + "     aurora_db_instance_identifier() di\n"
+      + "WHERE rt.server_id = di;";
+
```
Comment (davecramer) [inline wrapper/src/main/java/software/amazon/jdbc/plugin/staledns/AuroraStaleDnsHelper.java:58]: `select pg_is_in_recovery () as is_reader;` is much simpler

Comment (karenc-bq) [reply wrapper/src/main/java/software/amazon/jdbc/plugin/staledns/AuroraStaleDnsHelper.java:52]: Thank you

File: wrapper/src/main/java/software/amazon/jdbc/plugin/staledns/AuroraStaleDnsPlugin.java [hunk 6]
```diff
@@ -56,6 +56,8 @@ public class AuroraStaleDnsPlugin extends AbstractConnectionPlugin {
   private final PluginService pluginService;
   private final AuroraStaleDnsHelper helper;
 
+  static final String IS_READER_COLUMN = "is_reader";
```
Comment (sergiyvamz) [inline wrapper/src/main/java/software/amazon/jdbc/plugin/staledns/AuroraStaleDnsPlugin.java:59]: Do we still need it here in the  plugin? Should we relocate it to the helper class?

**Review by sergiyvamz — APPROVED** [review]

**Comment by shunyy** [comment]
@karenc-bq Hi!

I have run some tests on my end with snapshots and it seems to be working great!
Do you have any plans to release these changes?

**Comment by karenc-bq** [comment]
Hi @shunyy, thank you for testing out the changes.

Are you currently evaluating the driver for production use or do you have a timeline where you'd want an official release? We are aiming to do a aiming to do a release by end of January as there are other fixes that we want to include. https://github.com/awslabs/aws-advanced-jdbc-wrapper/pull/298. 

For convenience, we do offer snapshot builds here that would contains the fix from this PR: https://github.com/awslabs/aws-advanced-jdbc-wrapper/blob/main/docs/using-the-jdbc-driver/UsingTheJdbcDriver.md#using-a-snapshot-of-the-driver

**Comment by shunyy** [comment]
@karenc-bq 

Thanks for the reply!

I've been testing the snapshot version and wanted to use it in a production environment as soon as possible , but with the schedule you gave me, it's perfectly fine!

Looking forward to it!

**Pull request merged**